_posts/2019-08-01-my-own-summer-neural-network.md

```diff
@@ -0,0 +1,321 @@
+
+To get the current state of the keyboard, I used [pynput](https://pypi.org/project/pynput/) and came up with a script containing ugly global state :p
+
+```python
+
+from pynput import keyboard
+
+a_pressed = 0
+s_pressed = 0
+
+def on_press(key):
+    global a_pressed, s_pressed
+    if(key.char == 'a'):
+        a_pressed = 1
+    if(key.char == 's'):
+        s_pressed = 1
+
+def on_release(key):
+    global a_pressed, s_pressed
+    if(key.char == 'a'):
+        a_pressed = 0
+    if(key.char == 's'):
+        s_pressed = 0
+
+def kb_state():
+    return [a_pressed, s_pressed]
+
+listener = keyboard.Listener(
+    on_press=on_press,
+    on_release=on_release)
+listener.start()
+```
+
+Then I used this beast to capture the data and to store it to numpy file
+
+```python
+#!/usr/bin/env python3
+
+import numpy as np
+from PIL import ImageGrab
+import cv2
+import time
+import keypress
+import os
+
+file_name = 'training_data.npy'
+
+if(os.path.isfile(file_name)):
+    print('File exists')
+    training_data = list(np.load(file_name, allow_pickle=True))
+else:
+    print('File does not exists')
+    training_data = []
+
+
+while(True):
+    screen = np.array(ImageGrab.grab(bbox=(20,220, 420, 450)))
+
+    screen = cv2.resize(cv2.cvtColor(screen, cv2.COLOR_RGBA2GRAY), (80, 60))
+    keys = keypress.kb_state()
+    training_data.append([screen, keys])
+
+    if len(training_data) % 100 == 0:
+        print(len(training_data))
+        np.save(file_name, training_data)
+```
+
+What it does is:
+
+1. Take a screenshot
+2. Convert to grayscale
+3. Resize to 80x60
+4. Get the current keyboard state
+5. Store image and the corresponding keyboard state in an array of tuples
+
+The data is structured as:
+
+```
+[
+  [image_matrix, keyboard],
+  [image_matrix, keyboard],
+  ...
+  [image_matrix, keyboard]
+]
+```
+
+![Training data replayed](/img/endless-flighter-training-data.gif)
+
+Above is a visualisation of the learning data replayed. On the terminal, you can see which buttons were pressed, and on the window is the corresponding screen.
+
+I got pretty low frame rate of couple of images per second, but it turned out that this frame rate is actually enough for my simple Endless Flighter game. Note that images in learning data are cropped so that the bars behind player are not visible. This originates from my assumption, that the bars already passed do not matter anymore. When you play the game, you glaze your vision to the upper part of the screen.
+
+### 2. Capture Lots of Data by Flying Manually
+
+I wanted to make this work end-to-end with minimal training data set, because I did not want to spent hours flying the most dull game in the history for hours and then see that it didn't work and the data is corrupt somehow. So I decided just to fly for a couple of minutes and try to achieve end-to-end solution which will probably be really bad because of the small amount of data.
```

Comment:
```suggestion
I wanted to make this work end-to-end with a minimal training data set, because I did not want to spent hours flying the most dull game in the and then see that it didn't work and the data is corrupt somehow. So I decided just to fly for a couple of minutes and try to achieve end-to-end solution which will probably be really bad because of the small amount of data.
```